hal/src/thumbv7em/pwm.rs

```diff
@@ -12,76 +15,85 @@ use crate::target_device::{TC6, TC7};
 
 // Timer/Counter (TCx)
 
-pub enum TC0Pinout {
-    Pa5(Pa5<PfE>),
-    Pa9(Pa9<PfE>),
-    #[cfg(feature = "min-samd51j")]
-    Pb31(Pb31<PfE>),
-}
-
-pub enum TC1Pinout {
-    Pa7(Pa7<PfE>),
-    Pa11(Pa11<PfE>),
-}
-
-pub enum TC2Pinout {
-    Pa1(Pa1<PfE>),
-    Pa13(Pa13<PfE>),
-    Pa17(Pa17<PfE>),
-}
+/// This is a major syntax hack.
+///
+/// The previous Pinout types were enums that took specific v1::Pin types. As a
+/// result, there was no way to make that implementation simultaneously
+/// compatible with both v1::Pin and v2::Pin.
+///
+/// BUT, the enum variant syntax is the same as the namespaced function syntax.
+/// I converted the enums to structs, and I created constructor methods with the
+/// same names as the previous enum variants. By constructing Pinout types with
+/// functions rather than enum variants, you can make it generic over v1::Pin
+/// and v2::Pin types.
+///
+/// This is (mostly) backwards compatible with the current syntax, and all the
+/// existing calls compile. The only incompatible change is the requirement of
+/// type parameters on the Pwm types. Most of the type, the type parameters can
+/// be inferred, so this is mostly backwards compatible as well. But there were
+/// one or two instances where I had to add explicit type parameters to existing
+/// BSP code.
```

Comment:
I would just like to draw attention to this. It's very much not ideal, but I didn't see any other way to be simultaneously compatible with `v1` and `v2`.